### ISSUE TITLE
icon: a11y support appears broken

### ISSUE THREAD

**Issue by ThomasBurleson** [angular/material#9629]
Please see review here:

Aria-label has no effect.

``` js
    var label = attr.alt || attr.mdFontIcon || attr.mdSvgIcon || element.text();
    if ( !attr['aria-label'] ) {
    if (label !== '' && !parentsHaveText() ) {
```

Both aria-label and label are empty.
1. The logic is flawed.
   
   `if (label !== '' && !parentsHaveText() ) {`

Just because the parents have text does not mean that that text is going to be announced. It needs to be checked that the parent is a control or some kind of element such that the icon is not relevant.

**Comment by ThomasBurleson** [comment]
@jelbourn - FYI

**Comment by jelbourn** [comment]
CC @nbtorcol

**Comment by nbtorcol** [comment]
Correction, label does work, but aria-label still does not.

**Comment by ThomasBurleson** [comment]
@nbtorcol - can you provide more details and a specific use case where the problem manifests.

**Comment by nbtorcol** [comment]
A specific use case in what? The demo? I do not know what other details you
want. The two main points are that:

1.       The aria-label attribute is not getting filled

2.       The assumptions about parent text are incorrect

If you need more clarification on either of these, I need specific
questions.

**Comment by topherfangio** [comment]
@nbtorcol Are you saying that an existing demo is broken? If so, can you point us to which one?

If not, having a small Codepen, or even a code snippet of when your points 1 or 2 above actually break would be really helpful. We just want to make sure that we are solving your problem.

**Comment by nbtorcol** [comment]
The code is broken. For all of the icon demos that use aria-label, the
aria-label attribute is not getting processed.

    if ( !attr['aria-label'] ) {

is always getting evaluated.

Also, as I believe I wrote before:

1.       role=”img” should also be for the case where aria-label does exist

2.       The parentsHaveText function is not specific enough.

**Comment by ThomasBurleson** [comment]
@clshortfuse - can you work on this one. Thx.

**Comment by nbtorcol** [comment]
Actually, wait just a little longer. I am running one more test. I will
send an update as soon as I confirm.

**Comment by nbtorcol** [comment]
Confirmed. In demoLoadSvgIconsFromUrl, there is an icon with aria-label
cake, but the aria-label does not get processed. The part where label gets
assigned to aria-label along with role=”img” also gets dropped because
parentsHaveText does not work correctly. Finally, even if the aria-label
from the very first part did work right, it would not get processed by
browsers/screen readers because there is no role.

**Comment by nbtorcol** [comment]
I do not know how to fix aria-label not getting processed, but, if it did,
this is similar to what the code should look like

https://drive.google.com/a/google.com/file/d/0B0TBO_k0ZpHoNjhlWk9Pc0NBTE0/view?usp=sharing

**Comment by clshortfuse** [comment]
I've gone ahead and inspected the mdIconDirective code and looked at the ARIA guidelines.

Every `md-icon` should have one of the following: `aria-label`, `aria-labelledby`, `aria-describedby`

At the very least, if any of these 3 attributes are found, no changes should be made to the element. Choosing how to add these automatically is tricky. As stated before, just because the parent has text doesn't mean it complies with ARIA. Expanding on that point, just because a parent is labelled, it doesn't mean an mdIcon using font ligature should not include the `aria-hidden` attribute.

@nbtorcol I'm not sure the issue is about us automatically adding ARIA attributes. That can be prone to inaccurate implementation and it may also be computationally expensive if we decide to support `aria-labelledby`.

Our implementation isn't properly checking against ARIA spec, which means it's not alerting developers/designers of changes they need to make to the HTML.
